File: src/solutions/2023/06/go/main.go
package main

import (
	core "advent/of/go/src/langs/go"
	"math"
	"os"
	"strconv"
	"strings"
)

func main() {
	options := core.SolutionOptions{
		HasAlternate: false,
		HasIo:        false,
	}
	core.Run(os.Args, part1, part2, options)
}

func ohnomath(b float64, c float64) int64 {
	pepsilon := 0.001
	min := math.Floor((b+math.Sqrt(b*b-4*c))/2 - pepsilon)
	max := math.Ceil((b-math.Sqrt(b*b-4*c))/2 + pepsilon)
	return int64(min - max + 1)
}

func part1(input string, isTest bool) string {
	var td [2][]float64 = [2][]float64{}
	for i, line := range strings.Split(input, "\n") {
		ary := make([]float64, 0)
		for _, v := range strings.Split(strings.Split(line, ":")[1], " ") {
			if len(v) == 0 {
				continue
			}
			val, _ := strconv.ParseFloat(v, 64)
			ary = append(ary, val)
		}
		td[i] = ary
	}

	var res int64 = 1
	for i := 0; i < len(td[0]); i++ {
		res *= ohnomath(td[0][i], td[1][i])
	}

	return strconv.FormatInt(res, 10)
}

func part2(input string, isTest bool) string {
	var td [2]float64
	for i, line := range strings.Split(input, "\n") {
		val, _ := strconv.ParseFloat(strings.Join(strings.Split(strings.Split(line, ":")[1], " "), ""), 64)
		td[i] = val
	}

	return strconv.FormatInt(ohnomath(td[0], td[1]), 10)
}

File: src/solutions/2023/06/lua/main.lua
local run = require('src.langs.lua.run')
local str = require('src.langs.lua.utils.str')

local options = {
   has_alternate = false,
   has_io = false,
}

local function ohnomath(b, c)
   local pepsilon = 0.001
   local mn = math.floor((b + math.sqrt(b * b - 4 * c)) / 2 - pepsilon)
   local mx = math.ceil((b - math.sqrt(b * b - 4 * c)) / 2 + pepsilon)
   return mn - mx + 1
end

---@param input string
---@param is_test boolean
---@return string
local function part_1(input, is_test)
   local lines = str.split(input, '\n')

   local td = {}
   for i = 1, #lines do
      local line = lines[i]
      local ary = {}
      for _, v in ipairs(str.split(str.split(line, ':')[2], ' ')) do
         if #v > 0 then
            ary[#ary + 1] = tonumber(v)
         end
      end
      td[#td + 1] = ary
   end
   
   local res = 1
   for i = 1, #td[1] do
      res = res * ohnomath(td[1][i], td[2][i])
   end

   return tostring(res)
end

---@param input string
---@param is_test boolean
---@return string
local function part_2(input, is_test)
   local lines = str.split(input, '\n')

   local td = {}
   for i = 1, #lines do
      td[#td + 1] = tonumber(table.concat(str.split(str.split(lines[i], ':')[2], ' '), ''))
   end
   
   return tostring(ohnomath(td[1], td[2]))
end

local function main()
   run(arg, part_1, part_2, options)
end

main()
